File: app/notices.py
"""Gemeinsames Vokabular fuer Warnungen und Fehler.

Warnungen halten den Ablauf nicht auf, sie erscheinen in der UI und in der
PDF-Fusszeile. Fehler brechen ab. Beide tragen einen Code und benannte Parameter -
KEINEN fertigen Satz. Der Satz entsteht erst am Rand (HTTP-Antwort, PDF) aus dem
Sprachkatalog, sodass dieselbe Warnung deutsch oder englisch herauskommt, ohne dass
die Rechenschritte davon etwas wissen.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from app import config
from app.i18n import plain_params, translate


@dataclass(frozen=True)
class Notice:
    """Eine Warnung oder ein Hinweis zu einem Rechenergebnis."""

    code: str
    params: dict[str, object] = field(default_factory=dict)
    severity: str = "warn"  # "info" | "warn"

    def message(self, locale: str = config.DEFAULT_LOCALE) -> str:
        return translate(f"notices.{self.code}", locale, **self.params)


class AppError(Exception):
    """Abbruchfehler mit Code, Parametern und optionalem Feldbezug fuer die UI."""

    def __init__(self, code: str, field_name: str | None = None, **params: object) -> None:
        super().__init__(code)
        self.code = code
        self.field_name = field_name
        self.params = params

    def message(self, locale: str = config.DEFAULT_LOCALE) -> str:
        return translate(f"errors.{self.code}", locale, **self.params)


@dataclass
class NoticeList:
    """Sammler, damit Rechenschritte Warnungen anhaengen koennen, ohne sie zu kennen."""

    items: list[Notice] = field(default_factory=list)

    def warn(self, code: str, **params: object) -> None:
        self.items.append(Notice(code, params, "warn"))

    def info(self, code: str, **params: object) -> None:
        self.items.append(Notice(code, params, "info"))

    def extend(self, other: "NoticeList") -> None:
        self.items.extend(other.items)

    def as_dicts(self, locale: str = config.DEFAULT_LOCALE) -> list[dict[str, object]]:
        """Fuer die HTTP-Antwort. Der fertige Satz bleibt drin, damit /api/docs und
        jeder Nicht-Browser-Verbraucher lesbar bleiben - gerendert aus demselben
        Katalog, den auch die Oberflaeche benutzt."""
        return [
            {
                "code": notice.code,
                "params": plain_params(notice.params, locale),
                "severity": notice.severity,
                "message": notice.message(locale),
            }
            for notice in self.items
        ]

File: web/vision/notices.js
/**
 * notices.js - Vokabular fuer Warnungen und Abbrueche, wie app/notices.py.
 *
 * Code und benannte Parameter, KEIN fertiger Satz (AGENTS.md, Invariante 7).
 * Der Satz entsteht erst am Rand: die Oberflaeche uebersetzt `errors.<code>`
 * und `notices.<code>` aus demselben Katalog, den auch Python liest.
 *
 * Warum das hier trotzdem eine `message` mitgibt: die Antwortform ist die des
 * Servers, und der schickt sie mit (app/main.py::handle_app_error). Die
 * Oberflaeche zieht ihren eigenen Katalog vor und benutzt `message` nur als
 * Ausweg fuer einen Code, den sie nicht kennt - also darf sie nicht fehlen,
 * sonst saehe ein unbekannter Code im Ortsbetrieb anders aus als am Server.
 */

/** Ein Abbruch mit Code, Feldbezug und Parametern - das Gegenstueck zu AppError. */
export class AppError extends Error {
    constructor(code, fieldName = null, params = {}) {
        super(code);
        this.name = "AppError";
        this.code = code;
        this.field = fieldName;
        this.params = params;
    }

    /** Die Drahtform, die api.js::describeError erwartet. */
    toPayload(translate) {
        return {
            code: this.code,
            params: { ...this.params },
            field: this.field,
            message: translate(`errors.${this.code}`, this.params),
        };
    }
}

/** Sammler, damit Rechenschritte Warnungen anhaengen koennen, ohne sie zu kennen. */
export class NoticeList {
    constructor() {
        this.items = [];
    }

    warn(code, params = {}) {
        this.items.push({ code, params, severity: "warn" });
    }

    info(code, params = {}) {
        this.items.push({ code, params, severity: "info" });
    }

    asDicts(translate) {
        return this.items.map((notice) => ({
            code: notice.code,
            params: { ...notice.params },
            severity: notice.severity,
            message: translate(`notices.${notice.code}`, notice.params),
        }));
    }
}
